zml/module.zig

```diff
@@ -674,56 +765,74 @@ pub fn ExeWithWeights(comptime func: anytype) type {
         /// The raw untyped compiled module.
         inner: BaseExe,
 
-        /// The allocator used for bookkeeping.
-        allocator: std.mem.Allocator,
-
         /// Pre-allocated slice of buffers to use as inputs when the module is called.
-        input_buffers: []*pjrt.Buffer,
+        input_buffers: []const [*]*pjrt.Buffer,
 
         /// Pre-allocated slice of buffers to use as outputs when the module is called.
-        output_buffers: []*pjrt.Buffer,
+        output_buffers: []const [*]*pjrt.Buffer,
+
+        /// Internal memory slice used.
+        _data_buffer: []*pjrt.Buffer,
+
+        /// And the allocator backing _data_buffer.
+        _allocator: std.mem.Allocator,
 
         pub fn initFromModel(allocator: std.mem.Allocator, inner: BaseExe, model: Bufferized(Signature.ModelT)) !Self {
-            const input_buffers = try allocator.alloc(*pjrt.Buffer, inner.model_buffer_count + inner.args_buffer_count);
-            errdefer allocator.free(input_buffers);
-            fillBuffers(&model, input_buffers[0..inner.model_buffer_count]);
+            const n_buffers = inner.model_buffer_count + inner.args_buffer_count + inner.result_buffer_count;
+
+            const n_devices = inner.num_devices;
+            const n_ptrs = (2 + n_buffers) * inner.num_devices;
 
-            const output_buffers = try allocator.alloc(*pjrt.Buffer, inner.result_buffer_count);
-            errdefer allocator.free(output_buffers);
+            const data_buffer = try allocator.alloc(*pjrt.Buffer, n_ptrs);
+            errdefer allocator.free(data_buffer);
+
+            var fba = std.heap.FixedBufferAllocator.init(std.mem.sliceAsBytes(data_buffer));
+            const alloc = fba.allocator();
+
+            const input_buffers = alloc.alloc([*]*pjrt.Buffer, n_devices) catch unreachable;
+            for (input_buffers) |*input| {
+                input.* = (alloc.alloc(*pjrt.Buffer, inner.model_buffer_count + inner.args_buffer_count) catch unreachable).ptr;
+            }
+            fillBuffers(&model, input_buffers, 0, inner.model_buffer_count);
+
+            const output_buffers = alloc.alloc([*]*pjrt.Buffer, n_devices) catch unreachable;
+            for (output_buffers) |*output| {
+                output.* = (alloc.alloc(*pjrt.Buffer, inner.result_buffer_count) catch unreachable).ptr;
+            }
```

Comment:
Can you make sure we handle these `unreachable` instead ?

Since we are handling the error when allocating `data_buffer`

Reply:
those are provably unreachable because I've already made the actual allocation line 786.
this is just splitting the pre-existing alloc into smaller slices.
I thought it would be more idiomatic than using pointer arithmetic, but I can also do that if you prefer.

Reply:
I see. I would probably have used slicing because there is no actual allocations, but that's up to you

Reply:
ok I've rewrote this, let me know what you think